theories/datatypes/MSet.ec

```diff
@@ -0,0 +1,689 @@
+congr; apply perm_eqP_pred1.
+exact oflistK.
+qed.
+
+(* -------------------------------------------------------------------- *)
+abbrev disjoint (xs ys : 'a mset) = xs `&` ys = mset0.
+
+lemma disjointP (xs ys : 'a mset):
+  disjoint xs ys <=> forall (x : 'a), x \in xs => ! x \in ys.
+proof.
+split=> [disj_xs_ys x /mult_in_ge1 x_in_xs | all_xs_not_in_ys].
+- have := mset_eqP (xs `&` ys) mset0.
+  rewrite mult_in_ge1 {1}disj_xs_ys /= => inI_eq0 {disj_xs_ys}.
+  have:= inI_eq0 x; clear inI_eq0.  
+  rewrite !multE /#.
+rewrite mset_eqP => x.
+have := all_xs_not_in_ys x.
+rewrite !multE !mult_in_ge1.
+smt(mult_ge0).
+qed.
+
+(* -------------------------------------------------------------------- *)
+op [opaque] pick ['a] (A : 'a mset) = head witness (elems A).
+
+lemma pick0: pick<:'a> mset0 = witness.
+proof. by rewrite /pick elems_mset0. qed.
+
+lemma mem_pick (A : 'a mset): A <> mset0 => pick A \in A.
+proof.
+move=> /(contra _ _ (elems_eq_mset0 A)); rewrite /pick memE.
+by move=> /(mem_head_behead witness) <-.
+qed.
+
+(* -------------------------------------------------------------------- *)
+op (\subset) (s1 s2 : 'a mset) = forall x, s1 .[x] <= s2.[x].
+op (\proper) (s1 s2 : 'a mset) = s1 \subset s2 /\ s1 <> s2.
+
+lemma diff_mult_subset (s1 s2 : 'a mset): s2 \subset s1 => 
+  forall x, (s1 `\` s2).[x] = s1.[x] - s2.[x].
+proof. move => s1_sub_s2 x; rewrite diff_mult => /#. qed.
+
+lemma subset_trans (s1 s2 s3 : 'a mset): 
+  s1 \subset s2 => s2 \subset s3 => s1 \subset s3.
+proof. move=> le1 le2 x; exact (lez_trans _ _ _ (le1 x) (le2 x)). qed.
+
+lemma properP (s1 s2 : 'a mset) :
+  s1 \proper s2 <=> s1 \subset s2 /\ exists x, s1.[x] < s2.[x].
+proof.
+rewrite /(\proper) &(andb_id2l) mset_eqP negb_forall /= => le_XY.
+apply/eqboolP/exists_eq => x /= /#.
+qed.
+
+(* -------------------------------------------------------------------- *)
+lemma eqEsubset (s1 s2 : 'a mset) : 
+  (s1 = s2) <=> (s1 \subset s2) /\ (s2 \subset s1).
+proof. rewrite mset_eqP => /#. qed.
+
+lemma subset_msetU_id (A B : 'a mset) :
+  A \subset B => A `|` B = B.
+proof. move=> A_le_B; apply/mset_eqP=> x; rewrite union_mult /#. qed.
+
+lemma subset_msetI_id (A B : 'a mset) :
+  B \subset A => A `&` B = B.
+proof. move=> A_le_B; apply/mset_eqP=> x; rewrite intersection_mult /#. qed.
+
+lemma subset_mset_add (A B : 'a mset) : A \subset A + B.
+proof. rewrite /(\subset) => x; rewrite !multE; smt(mult_ge0). qed.
+
+lemma subset_msetDadd (A B : 'a mset) : A \subset B => (B `\` A) + A = B.
+proof. 
+move => /diff_mult_subset H; apply mset_eqP => x.
+rewrite multE H /#.
+qed.
+
+(* -------------------------------------------------------------------- *)
+lemma mset_ind (p : 'a mset -> bool):
+  p mset0 =>
+  (forall x s, p s => p (insert x s)) =>
+  (forall s, p s).
+proof.
+move=> p0 pa s; rewrite -elemsK.
+elim (elems s)=> {s} [|x s ih].
+- by rewrite -/mset0.
+rewrite oflist_cons.
+by apply pa.
+qed.
+
+(* ------------------------------------------------------------------ *)
```

Comment:
Maybe we should use a different symbols than `s` for arbitrary `mset` variables? `m` is already used for monoids, however.